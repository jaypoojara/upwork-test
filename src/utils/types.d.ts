declare type RootState = ReturnType<typeof store.getState>;

declare type CommonActionType = {
  type: string;
  payload: any;
}
