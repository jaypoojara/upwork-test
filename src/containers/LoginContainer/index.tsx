import {
  Box,
  Button,
  Container,
  CssBaseline,
  TextField,
  Typography
} from '@mui/material';
import React, {
  useEffect,
  useState
} from 'react';
import { useIntl } from 'react-intl';
import {
  useDispatch,
  useSelector
} from 'react-redux';

type Error = {
  username: boolean;
  password: boolean;
}

export default function DemoContainer() {
  const intl = useIntl();
  const dispatch = useDispatch();
  const [errors, setErrors] = useState<Error>({
    username: false,
    password: false
  });

  useEffect(() => {

  }, [dispatch]);

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    console.log(event.currentTarget)
    const data = new FormData(event.currentTarget);
    setErrors({
      username: data.get('username') === '',
      password: data.get('password') === '',
    })
  };

  return (
    <div>
      <Container component="main" maxWidth="sm" sx={{
        height: '100vh',
        alignItems: 'center',
        display: 'flex',
      }}>
        <Box
          sx={{
            display: 'flex',
            flexDirection: 'column',
            alignItems: 'center',
            border: '1px solid #979797',
            borderRadius: 10,
            boxShadow: 2,
            pl: 5,
            pr: 5,
            pt: 3,
            pb: 3
          }}
        >
          <Typography component="h1" variant="h5">
            {intl.formatMessage({ id: 'sign_in' })}
          </Typography>
          <Box component="form" onSubmit={handleSubmit} noValidate sx={{ mt: 1 }}>
            <TextField
              margin="normal"
              required
              fullWidth
              id="username"
              onChange={() => setErrors({
                ...errors,
                username: false
              })}
              error={errors.username}
              label={intl.formatMessage({id: 'username'})}
              name="username"
              autoFocus
              helperText={errors.username ? intl.formatMessage({id: 'username_error'}) : null}
            />
            <TextField
              margin="normal"
              required
              fullWidth
              name="password"
              onChange={() => setErrors({
                ...errors,
                password: false
              })}
              error={errors.password}
              label={intl.formatMessage({id: 'password'})}
              type="password"
              id="password"
              helperText={errors.password ? intl.formatMessage({id: 'password_error'}) : null}
            />
            <Button
              type="submit"
              size={'medium'}
              variant="contained"
              sx={{ mt: 3, mb: 2, width: '50%' }}
            >
              {intl.formatMessage({ id: 'sign_in' })}
            </Button>
          </Box>
        </Box>
      </Container>
    </div>
  );
}
