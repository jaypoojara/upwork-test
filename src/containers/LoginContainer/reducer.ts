import produce from 'immer';
import { createActions } from 'reduxsauce';

export const { Types: homeContainerTypes, Creators: homeContainerCreators } = createActions({
  requestGetDemo: ['name'],
  successGetDemo: ['data'],
  failureGetDemo: ['error'],
});

type InitialStateType = {
  name: string | null;
  error: string | null;
  loading: boolean | null;
}

export const initialState: InitialStateType = { name: null, error: null, loading: null };

export const homeContainerReducer = (state = initialState, action: any) =>
  produce(state, (draft) => {
    switch (action.type) {
      case homeContainerTypes.REQUEST_GET_DEMO:
        draft.name = null;
        draft.loading = true;
        break;
      case homeContainerTypes.SUCCESS_GET_DEMO:
        draft.name = action.data;
        draft.loading = false;
        break;
      case homeContainerTypes.FAILURE_GET_DEMO:
        draft.name = null;
        draft.error = action.error;
        draft.loading = false;
        break;
      default:
        return state;
    }
  })
export default homeContainerReducer;
