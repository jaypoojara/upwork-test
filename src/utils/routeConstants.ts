declare type RouteConstants = {
  [key: string]: {
    route: string;
    isProtected: boolean;
  }
}

const routeConstants: RouteConstants = {
  home: {
    route: '/',
    isProtected: false,
  },
  dashboard: {
    route: '/dashboard',
    isProtected: true,
  },
  notFound: {
    route: '/no-found',
    isProtected: false,
  }
};

export default routeConstants;
