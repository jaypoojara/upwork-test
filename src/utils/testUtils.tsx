import { ThemeProvider } from '@mui/material';
import { render } from '@testing-library/react';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { IntlGlobalProvider } from '../components/IntlGlobalProvider';
import configureStore from '../configureStore';
import { LanguageProvider } from '../containers/LanguageProvider';
import {
  DEFAULT_LOCALE,
  translationMessages
} from '../i18n';
import theme from '../theme/index';

export const renderWithIntl = (children: JSX.Element) =>
  render(
    <IntlProvider locale={DEFAULT_LOCALE} messages={translationMessages[DEFAULT_LOCALE]}>
      <IntlGlobalProvider>{children}</IntlGlobalProvider>
    </IntlProvider>
  );


export const renderProvider = (children: JSX.Element) => {
  const { store } = configureStore({});
  return render(
    <Provider store={store}>
      <LanguageProvider messages={translationMessages} locale={DEFAULT_LOCALE}>
        <ThemeProvider
          theme={theme}
        >
          <BrowserRouter>{children}</BrowserRouter>
        </ThemeProvider>
      </LanguageProvider>
    </Provider>
  );
};
