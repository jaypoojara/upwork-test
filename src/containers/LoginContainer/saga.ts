import {put, call, takeLatest} from 'redux-saga/effects';
import { timeout } from '../../utils';
import {appContainerTypes, appContainerCreators} from '../../AppReducer';

type LoginCallType = {
  username: string;
  password: string;
}

const {
  REQUEST_USER_LOGIN
} = appContainerTypes;

export function* loginApiCall({ username, password  }: LoginCallType) {
  yield call(timeout, 2000);
  if (username === 'upworkTest' && password === '2022') {
    yield put(appContainerCreators.successUserLogin(username))
  } else {
    yield put(appContainerCreators.failureUserLogin('Wrong password or username, please try again'))
  }
}

export default function* loginContainerSaga() {
  // @ts-ignore
  yield takeLatest(REQUEST_USER_LOGIN, loginApiCall);
}
