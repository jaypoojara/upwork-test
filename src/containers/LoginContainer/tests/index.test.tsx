import {
  fireEvent,
  waitFor,
} from '@testing-library/react';
import React from 'react';
import { timeout } from '../../../utils';
import routeConstants from '../../../utils/routeConstants';
import {
  renderProvider,
} from '../../../utils/testUtils';
import LoginContainer from '../index';

describe('<LoginContainer /> tests', () => {
  it('should render and match the snapshot', () => {
    const {baseElement} = renderProvider(<LoginContainer />)
    expect(baseElement).toMatchSnapshot();
  });

  it('should show validation error if username or password field is empty', () => {
    const screen = renderProvider(<LoginContainer />)
    const signInButton = screen.getByTestId("sign-in-button")
    fireEvent.click(signInButton);
    const errorMessage = screen.getByText("Please enter valid username");
    expect(errorMessage).toBeDefined();
  });

  it('should show validation error if username or password is wrong', async () => {
    const screen = renderProvider(<LoginContainer />)
    const signInButton = screen.getByTestId("sign-in-button")
    const usernameInput = screen.getByTestId("username-input");
    const passwordInput = screen.getByTestId("password-input");
    fireEvent.change(usernameInput, { target: { value: "dummy" } })
    fireEvent.change(passwordInput, { target: { value: "2022" } })
    fireEvent.click(signInButton);
    await timeout(2000);
    await waitFor(() => {
      const errorMessage = screen.getByText("Wrong password or username, please try again");
      expect(errorMessage).toBeDefined();
    })
  });

  it('should redirect user to dashboard if username and password is correct', async () => {
    const screen = renderProvider(<LoginContainer />)
    const signInButton = screen.getByTestId("sign-in-button")
    const usernameInput = screen.getByTestId("username-input");
    const passwordInput = screen.getByTestId("password-input");
    fireEvent.change(usernameInput, { target: { value: "upworkTest" } })
    fireEvent.change(passwordInput, { target: { value: "2022" } })
    fireEvent.click(signInButton);
    await timeout(2000);
    expect(window.location.pathname).toBe(routeConstants.dashboard.route);
  });
});
