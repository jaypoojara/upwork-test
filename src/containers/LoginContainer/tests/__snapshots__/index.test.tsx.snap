// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginContainer /> tests should render and match the snapshot 1`] = `
<body>
  <div>
    <div>
      <main
        class="MuiContainer-root MuiContainer-maxWidthSm css-9a9ws8-MuiContainer-root"
      >
        <div
          class="MuiBox-root css-bisg7f"
        >
          <h1
            class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
          >
            Sign in
          </h1>
          <form
            class="MuiBox-root css-164r41r"
            novalidate=""
          >
            <div
              class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-focused Mui-required css-147ndwt-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="true"
                for="username"
                id="username-label"
              >
                Username
                <span
                  aria-hidden="true"
                  class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-dqj5rd-MuiFormLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl css-3uibot-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  data-testid="username-input"
                  id="username"
                  name="username"
                  required=""
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-1in441m"
                  >
                    <span>
                      Username
                       
                      *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-1ykbbsm-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="password"
                id="password-label"
              >
                Password
                <span
                  aria-hidden="true"
                  class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-dqj5rd-MuiFormLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-3uibot-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  data-testid="password-input"
                  id="password"
                  name="password"
                  required=""
                  type="password"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-1ftyaf0"
                  >
                    <span>
                      Password
                       
                      *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <p
              class="MuiFormHelperText-root css-w5l3sh-MuiFormHelperText-root"
            />
            <button
              class="MuiLoadingButton-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root css-1p5nl2l-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
              data-testid="sign-in-button"
              id=":r2:"
              tabindex="0"
              type="submit"
            >
              Sign in
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </form>
        </div>
      </main>
    </div>
  </div>
</body>
`;
